packages/js/src/subscriptions/types.ts

```diff
@@ -0,0 +1,69 @@
+import type { RulesLogic } from 'json-logic-js';
+import type { TopicSubscription } from './subscription';
+import { SubscriptionPreference } from './subscription-preference';
+
+export type WorkflowIdentifierOrId = string;
+
+export type WorkflowFilter = {
+  label?: string;
+  workflowId: WorkflowIdentifierOrId;
+};
+
+export type WorkflowGroupFilter = {
+  label: string;
+  filter:
+    | { workflowIds?: Array<WorkflowIdentifierOrId>; tags?: string[] }
+    | ((args: {
+        preferences: Array<SubscriptionPreference>;
+      }) => Array<{ label: string; preference: SubscriptionPreference }>);
+};
+
+export type PreferenceFilter = WorkflowIdentifierOrId | WorkflowFilter | WorkflowGroupFilter;
+
+export type SubscriptionWorkflowPreference = {
+  workflowId: WorkflowIdentifierOrId;
+  value?: boolean | RulesLogic;
+};
+
+export type SubscriptionGroupPreference = {
+  group: Array<SubscriptionWorkflowPreference>;
+};
+
+export type SubscriptionPreferences = SubscriptionWorkflowPreference | SubscriptionGroupPreference;
+
+export type ListSubscriptionsArgs = {
+  topicKey: string;
+};
+
+export type GetSubscriptionArgs = {
+  topicKey: string;
+  identifier?: string;
+};
+
+export type CreateSubscriptionArgs = {
+  topicKey: string;
+  identifier?: string;
+  filters: Array<PreferenceFilter>;
+};
```

Comment:
```suggestion
export type CreateSubscriptionArgs = {
  topicKey: string;
  identifier?: string;
  preferences: Array<WorkflowIdentifierOrId | WorkflowFilter | WorkflowGroupFilter>
};
```

<br/>
where 

```
export type WorkflowGroupFilter = {
  label: string;
  filter: type.....
  condition: RulesLogic
};
```



sorry for bringing this up so late, but here are the use cases:

we need to allow passing conditions on *create*.
if someone is building a custom UI where the user interacts with the editor first and only later clicks the ‘Subscribe’ button, we need to support saving that state on create.

we also need to allow passing conditions on *update*.
for example, if a user applies a composed filter like:

   ```ts
   filter = { workflows: ["workflow-1"], tags: ["alerts"] }
   ```

we need to store the condition separately for each workflow. that means storing an individual preference entry per workflow along with its condition. then, on every update, we need to upsert the new state accordingly.